### PR TITLE
For #12345: avoid throwing regular exceptions for reportFullyDrawn. 

### PR DESCRIPTION
**Pull request by mcomella**
This should minorly improve performance – maybe it's not worth it.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/ext/Activity.kt [hunk 1]
```diff
@@ -29,12 +33,35 @@ fun Activity.enterToImmersiveMode() {
...
 fun Activity.reportFullyDrawnSafe() {
+    // An exception is thrown when calling reportFullyDrawn on some Samsung devices. We could catch
+    // the exception but catching exceptions is slow so we disable the call on all Samsung devices
+    // except the ones we're using for performance testing.
+    //
+    // We were unable to identify the root cause of the Exception but suspect it's related to Samsung
+    // security features. See
+    // https://github.com/mozilla-mobile/fenix/issues/12345#issuecomment-655058864 for details.
+    //
+    // We briefly measured how slow the Exception code path is here: on a Galaxy S5, it's ~1ms, ~6x
```
Comment (BranescuMihai) [inline app/src/main/java/org/mozilla/fenix/ext/Activity.kt:44]: @mcomella this is the first time I've seen code trying to reduce start-up time with 1ms, I'm impressed!

Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/ext/Activity.kt:44]: 😓

**Comment by mcomella** [comment]
I opened another Sentry report with a similar stack trace and found it was the same cause. Apparently, there are several of them and they're different stack traces because they're running different versions of Android. Therefore, my patch won't work as written :\ I'll rewrite it.

**Comment by mcomella** [comment]
@MarcLeclair `reportFullyDrawn` will throw exceptions on Samsung devices sometimes. My original code just caught the exception but, since exceptions are slow, this means a perf hit on these devices (on my GS5, it was ~1ms, ~6x slower than a check to avoid calling `reportFullyDrawn` on these devices). However, my check doesn't work as written so we'd need a new check. I was thinking of adding a blocklist of all Samsung devices except for a small set we allowlist (i.e. the ones we use in perf testing). However, this also won't work because the S7 shows up in crash reports.

To solve this problem, I can only think of using a separate app launch to set a preference we'd read on startup to run `reportFullyDrawn` but then that also means performance hits and much more complexity. At this point, I don't think it's worth implementing a fix for the perf hit of throwing exceptions.

What do you think? Perhaps there's another solution that comes to mind for you?

**Comment by mcomella** [comment]
> What do you think? Perhaps there's another solution that comes to mind for you?

iirc, I spoke to Marc on Zoom and we decided that we didn't want to add the complexity.

**Pull request closed without merging**